### PR TITLE
Fatal exception fixed in DVC on macOS while wxDataViewModel::Cleared …

### PR DESCRIPTION
**Pull request by valid-ptr**
…call + editing item.

Here are some inverstigated backtraces:

```
  * frame #0: 0x00007ff8132eeca8 libc++abi.dylib`__cxa_throw
    frame #1: 0x00007ff8131d3c9b libobjc.A.dylib`objc_exception_throw + 302
    frame #2: 0x00007ff81352e434 CoreFoundation`_CFThrowFormattedException + 202
    frame #3: 0x00007ff813399347 CoreFoundation`-[__NSArrayM objectAtIndex:] + 169
    frame #4: 0x00000001005afa27 ev_studio`-[wxCocoaOutlineDataSource outlineView:child:ofItem:] + 327
    frame #5: 0x00007ff8161921fb AppKit`refreshRowEntryItemAndLevelInfo + 167
    frame #6: 0x00007ff8165a64c8 AppKit`-[NSTableView _setNewObjectValueFromCell:ifNotEqualTo:forTableColumn:row:] + 148
    frame #7: 0x00007ff8165b2afc AppKit`-[NSTableView textDidEndEditing:] + 713
    frame #8: 0x00000001005b2c6b ev_studio`-[wxCocoaOutlineView textDidEndEditing:] + 43
    frame #9: 0x00007ff8133ecf23 CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    frame #10: 0x00007ff81348a3f9 CoreFoundation`___CFXRegistrationPost_block_invoke + 49
    frame #11: 0x00007ff81348a376 CoreFoundation`_CFXRegistrationPost + 496
    frame #12: 0x00007ff8133be836 CoreFoundation`_CFXNotificationPost + 733
    frame #13: 0x00007ff8142061ae Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 82
    frame #14: 0x00007ff8160e7185 AppKit`-[NSTextView(NSSharing) resignFirstResponder] + 771
    frame #15: 0x00007ff815f4eb07 AppKit`-[NSWindow _realMakeFirstResponder:] + 179
    frame #16: 0x000000010058b1fd ev_studio`-[wxNSWindow makeFirstResponder:] + 125
    frame #17: 0x00007ff816007d82 AppKit`-[NSWindow endEditingFor:] + 270
    frame #18: 0x00007ff815eed8d8 AppKit`-[NSTableView _endMyEditing] + 106
    frame #19: 0x00007ff8165b4115 AppKit`-[NSTableView _endMyEditingAndRemainFirstResponder] + 278
    frame #20: 0x00007ff815f450de AppKit`-[NSOutlineView reloadItem:reloadChildren:] + 993
    frame #21: 0x00000001005b540e ev_studio`non-virtual thunk to wxCocoaDataViewControl::Remove(wxDataViewItem const&) + 62
    frame #22: 0x00000001005c050f ev_studio`wxOSXDataViewModelNotifier::ItemDeleted(wxDataViewItem const&, wxDataViewItem const&) + 79
    frame #23: 0x00000001004adf40 ev_studio`wxDataViewModel::ItemDeleted(wxDataViewItem const&, wxDataViewItem const&) + 64
```

```
	0   CoreFoundation                      0x00007ff813472f0b __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007ff8131d3b9d objc_exception_throw + 48
	2   CoreFoundation                      0x00007ff8134f7adb -[NSObject(NSObject) __retain_OA] + 0
	3   CoreFoundation                      0x00007ff8133d9db9 ___forwarding___ + 1406
	4   CoreFoundation                      0x00007ff8133d97a8 _CF_forwarding_prep_0 + 120
	5   ev_studio                           0x00000001005afdcc -[wxCocoaOutlineDataSource outlineView:objectValueForTableColumn:byItem:] + 108
	6   AppKit                              0x00007ff8160c860b -[NSTableView preparedCellAtColumn:row:] + 425
	7   AppKit                              0x00007ff8161711f0 -[NSOutlineView preparedCellAtColumn:row:] + 51
	8   AppKit                              0x00007ff8160c8358 -[NSTableView _drawContentsAtRow:column:withCellFrame:] + 42
	9   AppKit                              0x00007ff816171160 -[NSOutlineView _drawContentsAtRow:column:withCellFrame:] + 71
	10  AppKit                              0x00007ff8160c7fc4 -[NSTableView drawRow:clipRect:] + 1638
	11  AppKit                              0x00007ff8160c763d -[NSTableView drawRowIndexes:clipRect:] + 709
	12  AppKit                              0x00007ff81616f866 -[NSOutlineView drawRowIndexes:clipRect:] + 94
	13  AppKit                              0x00007ff8160523bf -[NSTableView drawRect:] + 1670
	14  ev_studio                           0x00000001005a468f _ZN17wxWidgetCocoaImpl8drawRectEPvP6NSViewS0_ + 863
	15  ev_studio                           0x00000001005a1496 _Z14wxOSX_drawRectP6NSViewP13objc_selector6CGRect + 86
	16  AppKit                              0x00007ff815f81073 _NSViewDrawRect + 121
	17  AppKit                              0x00007ff81669bdec -[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:] + 1832
	18  AppKit                              0x00007ff815f80771 -[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:] + 747
	19  AppKit                              0x00007ff815f803b0 -[NSView(NSLayerKitGlue) drawLayer:inContext:] + 290
	20  QuartzCore                          0x00007ff81a48e31e CABackingStoreUpdate_ + 613
	21  QuartzCore                          0x00007ff81a4f3415 ___ZN2CA5Layer8display_Ev_block_invoke + 53
	22  QuartzCore                          0x00007ff81a48d6b8 -[CALayer _display] + 2268
	23  AppKit                              0x00007ff815f801e6 -[_NSBackingLayer display] + 462
	24  AppKit                              0x00007ff815ef573d -[_NSViewBackingLayer display] + 554
	25  QuartzCore                          0x00007ff81a48c631 _ZN2CA5Layer17display_if_neededEPNS_11TransactionE + 867
	26  QuartzCore                          0x00007ff81a5dc723 _ZN2CA7Context18commit_transactionEPNS_11TransactionEdPd + 717
	27  QuartzCore                          0x00007ff81a46e292 _ZN2CA11Transaction6commitEv + 704
	28  AppKit                              0x00007ff815f91592 __62+[CATransaction(NSCATransaction) NS_setFlushesWithDisplayLink]_block_invoke + 285
	29  AppKit                              0x00007ff8166d81d4 ___NSRunLoopObserverCreateWithHandler_block_invoke + 41
	30  CoreFoundation                      0x00007ff8133f6cb7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	31  CoreFoundation                      0x00007ff8133f6b54 __CFRunLoopDoObservers + 543
	32  CoreFoundation                      0x00007ff8133f5fe7 __CFRunLoopRun + 841
	33  CoreFoundation                      0x00007ff8133f55dd CFRunLoopRunSpecific + 563
	34  HIToolbox                           0x00007ff81c0284f1 RunCurrentEventLoopInMode + 292
	35  HIToolbox                           0x00007ff81c028118 ReceiveNextEventCommon + 284
	36  HIToolbox                           0x00007ff81c027fe5 _BlockUntilNextEventMatchingListInModeWithFilter + 70
	37  AppKit                              0x00007ff815e24b4c _DPSNextEvent + 886
	38  AppKit                              0x00007ff815e231b8 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1411
	39  AppKit                              0x00007ff815e155a9 -[NSApplication run] + 586
```

### CODE REVIEW

**Comment by vadz** [comment]
Thanks, I'm going to merge this unless @csomor can hint at a better way of dealing with this problem because this one is definitely very ugly -- but as long as we use it in the other places, we should use it here as well.

File: src/osx/dataview_osx.cpp
Comment (csomor) [inline src/osx/dataview_osx.cpp:259]: wouldn't it be better to store the former state here using `IsDeleting` and restore this state after, instead of unconditionally seting it to false

**Review by csomor — APPROVED** [review]
I'd prefer to have a cleaner model situation, that never gets out of sync, but until then, I'm fine of course with this. 
@valid-ptr Thanks for your fix. Could you please describe with a sentence how you can reproduce the crash in a sample ? so that we could later test any alternative solutions ?
Thanks

File: src/osx/dataview_osx.cpp
Comment (vadz) [reply src/osx/dataview_osx.cpp:259]: This is consistent with the existing usage of this function. We/I can change all of them at once later (I'd also use a RAII helper instead of doing it manually and get rid of `noFailureFlag`).

**Comment by valid-ptr** [comment]
> Could you please describe with a sentence how you can reproduce the crash in a sample ? so that we could later test any alternative solutions

I've modified dataview sample: Just press Ctrl-W while editing any artist with and without prev. commit.
Without 1st commit: Sometimes it crashes immediately, sometimes - not (but soon).
With 1st commit: No crash

**Comment by vadz** [comment]
@csomor Should I merge this in the current state or do you plan to improve on this?

**Comment by csomor** [comment]
> @csomor Should I merge this in the current state or do you plan to improve on this?

@vadz please commit, I tried to understand the sequence tying native and wx-model, but I'll need way more time to get a grip on that :-)

**Comment by vadz** [comment]
Thanks again, @valid-ptr, finally merging this now after simplifying the code just a bit.

**Comment by vadz** [comment]
I don't fully understand what's going on here because somehow the original commit (8aae7ad9379289e770a72983d0c5151f1da763cc) did work on top of its original parent 32eb4af061 (Avoid using wxClientDC to redraw grid lines, 2022-01-17), but doesn't work now that it's merged in master nor if I just just cherry-pick it to recent master, e.g. 995c6e6df5 (Add wxSpinCtrl::SetIncrement() and implement it for all ports, 2022-01-15) which is the latest commit before the merge.

The mysterious part is that there don't seem to be any commits affecting wxOSX except for bitmap scaling-related ones, nor wxDVC, nor the test between 32eb4af061 and 995c6e6df5, but debugging it right now it seems clear that returning `nil` from `-[wxCocoaOutlineDataSource outlineView:child:ofItem:]` is wrong, as this really does break switching selection after deleting an item. I.e. the question is not why it doesn't work now, but rather why did it work before...

Anyhow, the unit test is fixed by the changes of ff8e60caeaed8772759d41b41a44c2e70a7068b4 and clearing the model in the sample doesn't result in anything bad happening for me in a build with ASAN, so I hope that it's fixed now, but please let me know if you see something wrong. I still have no idea why didn't the test fail before though.

**Pull request merged**
